### DISCUSSION TITLE
bug: Type error when calling .meta() with partial attributes

### DISCUSSION THREAD

**Discussion by tomagranate** [trpc/trpc#4300] [Q&A]
### Provide environment information

```
  System:
    OS: macOS 13.3.1
    CPU: (16) x64 Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
    Memory: 42.06 GB / 64.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 18.15.0 - ~/.nvm/versions/node/v18.15.0/bin/node
    npm: 9.6.4 - ~/.nvm/versions/node/v18.15.0/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  Browsers:
    Chrome: 112.0.5615.137
    Edge: 112.0.1722.68
    Firefox: 112.0.2
    Safari: 16.4
  npmPackages:
    @tanstack/react-query: ^4.29.5 => 4.29.5
    @trpc/client: ^10.23.0 => 10.23.0
    @trpc/next: ^10.23.0 => 10.23.0
    @trpc/react-query: ^10.23.0 => 10.23.0
    @trpc/server: ^10.23.0 => 10.23.0
    next: 13.3.1 => 13.3.1
    react: 18.2.0 => 18.2.0
    typescript: 5.0.4 => 5.0.4
```

### Describe the bug

Composing `.meta({ ... })` calls complain with a Type Error if you only pass some of the keys of your defined `Meta` type, even if all fields are defined properly with `defaultMeta`.

The actual composition of the fields works correctly! You can override only one attribute of your meta, and that does actually take effect. It seems to be just a type error.

e.g. 
```
interface Meta {
  att1: string;
  att2: string;
}

const t = initTRPC
  .meta<Meta>()
  .create(
    defaultMeta: {
      att1: 'hello',
      att2: 'world'
    },
  )

const procedure = t.procedure.meta({
  att1: 'override',
})
// ^^^ causes Type Error, because not all attributes are provided here
```

Example Error:
```
Argument of type '{ att1: string; }' is not assignable to parameter of type 'Meta'.
  Property 'att2' is missing in type '{ att1: string; }' but required in type 'Meta'.(2345)
```

### Link to reproduction

https://stackblitz.com/edit/github-dhd5yb-kanmz4?file=src/pages/api/trpc/[trpc].ts

### To reproduce

You can see in the Stackblitz provided, the type errors appear, but the actual behaviour is correct.

You can reproduce this directly with the [Default Meta example in the docs](https://trpc.io/docs/server/metadata#default-meta-chaining-and-shallow-merging) as well

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

**Comment by Nick-Lucas** [comment]
The trouble is Typescript has no way of knowing that you've provided the other keys elsewhere, so you need to mark them as optional if you want to merge a partial update. _Maybe_ there a way to make that work better? I agree it would be nicer if the procedure was aware of how Meta has already been stacked.

PRs even just with a proof of concept are welcome :)
